File: cpp/Hand.cpp
#include <iostream>
#include <vector>
#include <algorithm> // For std::sort
#include "Card.h"
#include "Hand.h"

using namespace std;

// Constructor that initializes the hand and evaluates its type
Hand::Hand(vector<Card> cards) {
    this->cards = cards; // Directly initialize the vector
    separateKeyCardsAndKickers(); // Separate key cards from kickers
    sortHand(); // Sort the cards based on hand value
    evaluateHand(); // Determine the type of hand and assign values
}

// Get the list of cards in the hand
vector<Card> Hand::getCards() const {
    return cards;
}

vector<Card> Hand::getUnsortedCards() const {
    return unsorted;
}

void Hand::setUnsorted(vector<Card> cardList) {
    this->unsorted = cardList; // Directly initialize the vector
}

// Get the type of the hand (e.g "pair", "three of a kind")
string Hand::getHandType() {
    return handType;
}

// Generate a string representation of the hand
string Hand::toString() const {
    string result = "";
    for (const Card& card : unsorted) {
        result += card.toString() + " "; // Format each card in the hand
    }
    return result;
}

std::ostream& operator<<(std::ostream& os, const Hand& hand){
    os << hand.toString(); // Use the toString method for output
    return os;
}

// Evaluate the type of hand (e.g., "flush", "straight", etc.)
void Hand::evaluateHand() {
    // Vector to count occurrences of each face value (index 0 is unused)
    vector<int> faceCounts(15, 0);
    for (const Card& card : cards) {
        faceCounts[(int)(card.faceValue())]++;
    }

    // Check if the hand is a flush 
    bool isFlush = this -> isFlush();
    // Check if the hand is a straight 
    bool isStraight = this -> isStraight();

    // Determine hand type and assign a base value to each card
    if (isFlush && isStraight) {
        int a = 0;
        if (faceCounts[14] > 0 && faceCounts[13] > 0) { // if it includes the Ace and King
            handType = "Royal Flush";
            a += 1000;
        } else {
            handType = "Straight Flush"; 
        }
        a += 9000;
        setHandValueToCards(a);
    } else if (containsNOfAKind(faceCounts, 4)) {
        handType = "Four of a Kind";
        setHandValueToCards(8000); 
    } else if (containsNOfAKind(faceCounts, 3) && containsNOfAKind(faceCounts, 2)) {
        handType = "Full House";
        setHandValueToCards(7000); 
    } else if (isFlush) {
        handType = "Flush";
        setHandValueToCards(6000); 
    } else if (isStraight) {
        handType = "Straight";
        setHandValueToCards(5000); 
    } else if (containsNOfAKind(faceCounts, 3)) {
        handType = "Three of a Kind";
        setHandValueToCards(4000); 
    } else if (countPairs(faceCounts) == 2) {
        handType = "Two Pair";
        setHandValueToCards(3000); 
    } else if (containsNOfAKind(faceCounts, 2)) {
        handType = "One Pair";
        setHandValueToCards(2000); 
    } else {
        handType = "High Card";
        setHandValueToCards(1000);
    }
}

void Hand::setHandValueToCards(double baseValue) {
    for (Card& card : cards) {
        double handValue;
        if (handType != "One Pair" && handType != "Two Pair") {
            handValue = baseValue + card.faceValue() + card.suitValue();
            //cout << baseValue << card.faceValue()<<card.suitValue()<<endl;
        } else {
            handValue = baseValue + card.faceValue();
        }
        card.setHandValue(handValue);
    }
}

// Check if all cards in the hand have the same suit
bool Hand::isFlush() {
    string suit = cards[0].getSuit();
    for (const Card& card : cards) {
        if (card.getSuit() != suit) {
            return false;
        }
    }
    return true;
}

// Check if the cards form a straight
bool Hand::isStraight() {
    vector<int> values;
    for (const Card& card : cards) {
        values.push_back(static_cast<int>(card.faceValue()));
    }
    sort(values.begin(), values.end());

    // Check for high Ace straight (10-J-Q-K-A)
    if (values[0] == 2 && values[4] == 14 && values[1] == 3 &&
        values[2] == 4 && values[3] == 5) {
        return true;
    }

    for (size_t i = 1; i < values.size(); i++) {
        if (values[i] != values[i - 1] + 1) {
            return false;
        }
    }
    return true;
}

// Check if there's a card value with exactly n occurrences
bool Hand::containsNOfAKind(const vector<int>& faceCounts, int n) {
    for (int count : faceCounts) {
        if (count == n) {
            return true;
        }
    }
    return false;
}

// Count the number of pairs in the hand
int Hand::countPairs(const vector<int>& faceCounts) {
    int pairs = 0;
    for (int count : faceCounts) {
        if (count == 2) {
            pairs++;
        }
    }
    return pairs;
}

// Separate cards into key cards (cards forming the hand) and kickers (remaining cards)
void Hand::separateKeyCardsAndKickers() {
    keyCards.clear();
    kickers.clear();

    vector<int> faceCounts(15, 0);
    for (const Card& card : cards) {
        faceCounts[static_cast<int>(card.faceValue())]++;
    }

    // Add key cards to keyCards and remaining cards to kickers
    for (const Card& card : cards) {
        if (faceCounts[static_cast<int>(card.faceValue())] > 1) {
            keyCards.push_back(card);
        } else {
            kickers.push_back(card);
        }
    }
}

// Sort key cards and kickers by their hand values
void Hand::sortHand() {
    // Sort key cards and kickers in descending order of face value
    sort(keyCards.begin(), keyCards.end(), [](const Card& a, const Card& b) {
        return b.faceValue() + b.suitValue()< a.faceValue() + b.suitValue();
    });
    sort(kickers.begin(), kickers.end(), [](const Card& a, const Card& b) {
        return b.faceValue() +b.suitValue()< a.faceValue()+a.suitValue();
    });

    // Combine key cards and kickers into the final sorted hand
    cards.clear();
    cards.insert(cards.end(), keyCards.begin(), keyCards.end()); // Add key cards first
    cards.insert(cards.end(), kickers.begin(), kickers.end());   // Add kickers after key cards

    // Adjust for high Ace straight
    if (isStraight() && static_cast<int>(cards[0].faceValue()) == 14 && 
        static_cast<int>(cards[4].faceValue()) == 2) {
        cards.push_back(cards[0]);
        cards.erase(cards.begin()); // Remove the first element
    }
}

// Compare this hand to another hand based on card values
double Hand::compareTo(Hand& other) {
    for (size_t i = 0; i < this-> keyCards.size(); i++) {
        double cardComparison = this->cards[i].getHandValue() - other.cards[i].getHandValue();
        
        if (cardComparison != 0) {
            return cardComparison;
        }
    }
    for (Card& card : cards) { // If the hands are the same, take suit into account
        card.setHandValue(card.suitValue());
    }
    for (size_t i = this -> keyCards.size(); i < cards.size(); i++) {
        double cardComparison = this->cards[i].getHandValue() - other.cards[i].getHandValue();
        
        if (cardComparison != 0) {
            return cardComparison;
        }
    }
    return 0; // Hands are equal if all card comparisons are equal
}

// Overload the < operator
    bool Hand::operator<(Hand& other){
        return compareTo(other) < 0; // Less than zero indicates this is less than 'other'
    }
    

File: python/Hand.py
from typing import List
from Card import *

class Hand:
    def __init__(self, cards: List['Card']):
        self.cards = list(cards)  # List of cards in the hand
        self.unsorted = []  # List of cards in the hand unsorted
        self.hand_type = ""  # Type of hand (e.g., "pair", "three of a kind", etc.)
        self.key_cards = []  # Cards that form the key part of the hand
        self.kickers = []  # Remaining cards in the hand
        self.separate_key_cards_and_kickers()  # Separate key cards from kickers
        self.evaluate_hand()  # Determine the type of hand and assign values
        self.sort_hand()  # Sort the cards based on hand value

    def get_cards(self):
        return self.cards

    def get_unsorted_cards(self):
        return self.unsorted

    def set_unsorted(self, card_list: List['Card']):
        self.unsorted = list(card_list)

    def get_hand_type(self):
        return self.hand_type

    def __str__(self):
        return ' '.join(str(card).rjust(3) for card in self.unsorted)  
        # Format each card in the hand

    def evaluate_hand(self):
        face_counts = [0] * 15  # Array to count occurrences of each face value
        for card in self.cards:
            face_counts[int(card.face_value())] += 1

        is_flush = self.is_flush()  # Check if the hand is a flush
        is_straight = self.is_straight()  # Check if the hand is a straight

        # Determine hand type and assign a base value to each card
        if is_flush and is_straight:
            if face_counts[14] > 0 and face_counts[13] > 0:  # Includes Ace and King
                self.hand_type = "Royal Flush"
                base_value = 10000
            else:
                self.hand_type = "Straight Flush"
                base_value = 9000
            self.set_hand_value_to_cards(base_value)
        elif self.contains_n_of_a_kind(face_counts, 4):
            self.hand_type = "Four of a Kind"
            self.set_hand_value_to_cards(8000)
        elif self.contains_n_of_a_kind(face_counts, 3) and self.contains_n_of_a_kind(face_counts, 2):
            self.hand_type = "Full House"
            self.set_hand_value_to_cards(7000)
        elif is_flush:
            self.hand_type = "Flush"
            self.set_hand_value_to_cards(6000)
        elif is_straight:
            self.hand_type = "Straight"
            self.set_hand_value_to_cards(5000)
        elif self.contains_n_of_a_kind(face_counts, 3):
            self.hand_type = "Three of a Kind"
            self.set_hand_value_to_cards(4000)
        elif self.count_pairs(face_counts) == 2:
            self.hand_type = "Two Pair"
            self.set_hand_value_to_cards(3000)
        elif self.contains_n_of_a_kind(face_counts, 2):
            self.hand_type = "One Pair"
            self.set_hand_value_to_cards(2000)
        else:
            self.hand_type = "High Card"
            self.set_hand_value_to_cards(1000)

    def set_hand_value_to_cards(self, base_value):
        for card in self.cards:
            if self.hand_type not in ["One Pair", "Two Pair"]:
                hand_value = base_value + card.face_value() + card.suit_value()
            else:
                hand_value = base_value + card.face_value()
            card.set_hand_value(hand_value)

    def is_flush(self):
        suit = self.cards[0].get_suit()
        return all(card.get_suit() == suit for card in self.cards)

    def is_straight(self):
        values = sorted(int(card.face_value()) for card in self.cards)
        
        # Check for high Ace straight (10-J-Q-K-A)
        if values[0] == 2 and values[4] == 14 and values[1] == 3 and values[2] == 4 and values[3] == 5:
            return True

        return all(values[i] == values[i - 1] + 1 for i in range(1, len(values)))

    def contains_n_of_a_kind(self, face_counts, n):
        return any(count == n for count in face_counts)

    def count_pairs(self, face_counts):
        return sum(1 for count in face_counts if count == 2)

    def separate_key_cards_and_kickers(self):
        self.key_cards = []
        self.kickers = []
        face_counts = [0] * 15

        for card in self.cards:
            face_counts[int(card.face_value())] += 1

        for card in self.cards:
            if face_counts[int(card.face_value())] > 1:
                self.key_cards.append(card)
            else:
                self.kickers.append(card)

    def sort_hand(self):
        self.key_cards.sort(key=lambda card: card.face_value()+card.suit_value(), reverse=True)
        self.kickers.sort(key=lambda card: card.face_value()+card.suit_value(), reverse=True)

        self.cards.clear()
        self.cards.extend(self.key_cards)  # Add key cards first (for other types)
        self.cards.extend(self.kickers)  # Add kickers after key cards

        if self.is_straight() and int(self.cards[0].face_value()) == 14 and int(self.cards[4].face_value()) == 2:
            self.cards.append(self.cards[0])
            self.cards.pop(0)

    def compare_to(self, other):
        for i in range(len(self.key_cards)):
            card_comparison = (self.cards[i].get_hand_value() > other.cards[i].get_hand_value()) - (self.cards[i].get_hand_value() < other.cards[i].get_hand_value())
            if card_comparison != 0:
                return card_comparison

        for card in self.cards:  # If the hands are the same, take suit into account
            card.set_hand_value(card.suit_value())

        for i in range(len(self.key_cards), len(self.cards)):
            card_comparison = (self.cards[i].get_hand_value() > other.cards[i].get_hand_value()) - (self.cards[i].get_hand_value() < other.cards[i].get_hand_value())
            if card_comparison != 0:
                return card_comparison

        return 0  # Hands are equal if all card comparisons are equal (impossible here)
    def __lt__(self, other):
        return self.compare_to(other) < 0